### PR TITLE
context_servers: Add initial implementation

### PR DESCRIPTION
**Pull request by dsp-ant**
This commit proposes the addition of "context serveres" and the
underlying protocol (model context protocol). Context servers allow
simple definition of slash commands in another language and running
local on the user machines. This aims to quickly prototype new commands,
and provide a way to add personal (or company wide) customizations to
the assistant panel, without having to maintain an extension. We can
use this to reuse our existing codebase, with authenticators, etc and
easily have it provide context into the assistant panel.

As such it occupies a different design space as extensions, which I think are
more aimed towards long-term, well maintained pieces of code that can be
easily distributed.

It's implemented as a central crate for easy reusability across the codebase
and to easily hook into the assistant panel at all points.

Design wise there are a few pieces:
1. client.rs: A simple JSON-RPC client talking over stdio to a spawned server. This is
   very close to how LSP work and likely there could be a combined client down the line.
2. types.rs: Serialization and deserialization client for the underlying model context protocol.
3. protocol.rs: Handling the session between client and server.
4. manager.rs: Manages settings and adding and deleting servers from a central pool.

A server can be defined in the settings.json as:

```
"context_servers": [
   {"id": "test", "executable": "python", "args": ["-m", "context_server"]
]
```

## Quick Example
A quick example of how a theoretical backend site can look like. With roughly 100 lines
of code (nicely generated by Claude) and a bit of decorator magic (200 lines in total), one
can come up with a framework that makes it as easy as:

```python
@context_server.slash_command(name="rot13", description="Perform a rot13 transformation")
@context_server.argument(name="input", type=str, help="String to rot13")
async def rot13(input: str) -> str:
    return ''.join(chr((ord(c) - 97 + 13) % 26 + 97) if c.isalpha() else c for c in echo.lower())
```

to define a new slash_command.

## Todo:
 - Allow context servers to be defined in workspace settings.
 - Allow passing env variables to context_servers

Release Notes:

- N/A

### CODE REVIEW

**Comment by dsp-ant** [comment]
I'll work through the issue with the RwLock held during await.

File: assets/settings/default.json [hunk 1]
```diff
@@ -997,5 +997,16 @@
...
+  "ssh_connections": null,
+  // Configures the Context Server Protocol binaries
+  //
+  // Examples:
+  // {
+  //   "id": "server-1",
+  //   "executable": "/path",
+  //   "args": ['arg1", "args2"]
+  // }
+  "experimental.context_servers": {
```
Comment (rtfeldman) [inline assets/settings/default.json:1009]:
```suggestion
  "experimental_context_servers": {
```

I like including "experimental in here" but I don't think we have a convention of using dots in key names - underscore seems fine to me! 😄

Comment (maxdeviant) [reply assets/settings/default.json:1009]: The dot is fine, it's what we do for `experimental.theme_overrides`.

Comment (rtfeldman) [reply assets/settings/default.json:1009]: TIL! 👍

**Review by rtfeldman — APPROVED** [review]
Overall approach looks good! I left a couple of comments, but nothing blocking.

Thanks for putting all this together @dsp-ant! 😃

File: crates/assistant/src/assistant.rs [hunk 2]
```diff
@@ -237,9 +241,67 @@ pub fn init(fs: Arc<dyn Fs>, client: Arc<Client>, cx: &mut AppContext) -> Arc<Pr
...
+            context_servers::manager::Event::ServerStarted { server_id } => {
+                cx.update_model(
+                    &manager,
+                    |manager: &mut context_servers::manager::ContextServerManager, cx| {
+                        let slash_command_registry = SlashCommandRegistry::global(cx);
+                        let cs_registry = context_server_command::ContextServerRegistry::global(cx);
+                        if let Some(server) = manager.get_server(server_id) {
+                            cx.spawn(|_, _| async move {
+                                if let Some(protocol) = server.client.read().as_ref() {
+                                    if let Ok(prompts) = protocol.list_prompts().await {
```
[the remark this answers is not in the document]
Comment (maxdeviant) [reply crates/assistant/src/assistant.rs]: Good idea! Done in 601093dbbe.

**Review by maxdeviant — APPROVED** [review]
I addressed the Clippy failure and did a pass on cleaning up a few things.

I think this is in a state that we can go ahead and merge. Thanks!

**Pull request merged**